### ISSUE TITLE
Evaluation metrics on pytorch tensors discrepancy with MONAI metrics

### ISSUE THREAD

**Issue by jakubMitura14** [rundherum/pymia#46]
Hello some time ago I told that I am working on CUDA accelerated metrics - works go slowly on, Hovewer - now I am working on benchmarking against your CPU and monai GPU - and using your and monai metrics I get diffrent results - probably it is my mistake - code is below Data is available under link below i Use volume and label 21 below

https://wiki.cancerimagingarchive.net/display/Public/CT-ORG%3A+CT+volumes+with+multiple+organ+segmentations

```
!python -c "import monai" || pip install -q "monai-weekly[gdown, nibabel, tqdm, ignite]"
!python -c "import matplotlib" || pip install -q matplotlib
```

```
from monai.utils import first, set_determinism
from monai.transforms import (
    AsDiscrete,
    AsDiscreted,
    EnsureChannelFirstd,
    Compose,
    CropForegroundd,
    LoadImaged,
    Orientationd,
    RandCropByPosNegLabeld,
    ScaleIntensityRanged,
    Spacingd,
    EnsureTyped,
    EnsureType,
    Invertd,
)
from monai.handlers.utils import from_engine
from monai.networks.nets import UNet
from monai.networks.layers import Norm
from monai.metrics import DiceMetric
from monai.losses import DiceLoss
from monai.inferers import sliding_window_inference
from monai.data import CacheDataset, DataLoader, Dataset, decollate_batch
from monai.config import print_config
from monai.apps import download_and_extract
import torch
import matplotlib.pyplot as plt
import tempfile
import shutil
import os
import glob



from monai.transforms import (
    LoadImage, LoadImaged, EnsureChannelFirstd,
    Resized, EnsureTyped, Compose
)
from monai.config import print_config

filename = f"C:\\Users\\1\\Downloads\\volumes 0-49\\volumes 0-49\\volume-21.nii.gz"
biggestDat, meta = LoadImage()(filename)
labelFilename = f"C:\\Users\\1\\Downloads\\volumes 0-49\\labels\\labels-21.nii.gz"
biggestDatLabel, metaLabel = LoadImage()(labelFilename)


imageTensor = torch.Tensor(biggestDat)
imageLabel = torch.Tensor(biggestDatLabel)
#preparing some tensors for later comparison
onlyBladder =torch.where( imageLabel==2, 1, 0)
onlyBrain =torch.where( imageLabel==6, 1, 0)
onlyBone =torch.where( imageLabel==5, 1, 0)
sizzz = onlyBone.size()

not_translated = onlyBone.clone().to(torch.device("cuda"))

translated = torch.zeros((sizzz)).to(torch.device("cuda"))
translated[5:512,5:512,5:826]= onlyBone.clone()[0:507,0:507,0:821]

dice_metric = DiceMetric(include_background=True, reduction="mean")
dice_metric(y_pred=not_translated, y=translated)  # callable to add metric to the buffer
metric = dice_metric.aggregate().item()

#### monai metric will evaluate to 0.532331645488739
### now PYMIA

import glob
import os

import numpy as np
import pymia.evaluation.metric as metric
import pymia.evaluation.evaluator as eval_
import pymia.evaluation.writer as writer
import SimpleITK as sitk

metrics = [metric.DiceCoefficient()]

labels = {1: 'lab'
          }
evaluator = eval_.SegmentationEvaluator(metrics, labels)

evaluator.evaluate(not_translated.cpu() , translated.cpu(), "1")
writer.ConsoleWriter().write(evaluator.results)
## PYMIA evaluates to 0.757



```

Thanks For Help

**Comment by fabianbalsiger** [comment]
Hi Jakub,

well if you look at the [MONAI documentation](https://docs.monai.io/en/latest/metrics.html#mean-dice), you will see that the input format is quite different from what you are doing. So, that is probably where the difference origins from. As a quick check, I used scipy, which evaluates to 0.757 as pymia:

```
from scipy.spatial import distance
dice = 1 - distance.dice(not_translated.cpu().numpy().flatten() , translated.long().cpu().numpy().flatten())
```

So, I think pymia is returning the correct Dice coefficient. Feel free to reopen if you have further questions.

**Comment by jakubMitura14** [comment]
Ypu must bu right , thanks !
